src/main/java/net/neoforged/neoforge/attachment/AttachmentType.java

```diff
@@ -75,21 +76,50 @@ private static <T> IAttachmentComparator<T> defaultComparator(IAttachmentSeriali
     /**
      * Creates a builder for an attachment type.
      *
+     * <p>See {@link #builder(Function)} for attachments that want to capture a reference to their holder.
+     *
      * @param defaultValueSupplier A supplier for a new default value of this attachment type.
      */
     public static <T> Builder<T> builder(Supplier<T> defaultValueSupplier) {
+        return builder(holder -> defaultValueSupplier.get());
+    }
+
+    /**
+     * Creates a builder for an attachment type.
+     *
+     * <p>This overload allows capturing a reference to the {@link IAttachmentHolder} for the attachment.
+     * To obtain a specific subtype, the holder can be cast.
+     * See {@link #builder(Supplier)} for an overload that does not capture the holder.
+     *
+     * @param defaultValueSupplier A supplier for a new default value of this attachment type.
+     */
+    public static <T> Builder<T> builder(Function<IAttachmentHolder, T> defaultValueSupplier) {
```

Comment:
I would rename this parameter as `supplier` might be confusing when it's a function, and there's also another variant with a supplier. Same below.